relay-sampling/src/dsc.rs

```diff
@@ -145,17 +145,27 @@ mod sample_rate_as_string {
 
     use serde::{Deserialize, Serialize};
 
+    #[derive(Debug, Clone)]
+    enum StringOrFloat<'a> {
+        String(Cow<'a, str>),
+        Float(f64),
+    }
+
     pub fn deserialize<'de, D>(deserializer: D) -> Result<Option<f64>, D::Error>
     where
         D: serde::Deserializer<'de>,
     {
-        let value = match Option::<Cow<'_, str>>::deserialize(deserializer)? {
+        let value = match Option::<StringOrFloat>::deserialize(deserializer)? {
             Some(value) => value,
             None => return Ok(None),
         };
 
-        let parsed_value =
-            serde_json::from_str(&value).map_err(|e| serde::de::Error::custom(e.to_string()))?;
+        let parsed_value = match value {
+            StringOrFloat::Float(f) => f,
+            StringOrFloat::String(s) => {
+                serde_json::from_str(&s).map_err(|e| serde::de::Error::custom(e.to_string()))?
+            }
+        };
```

Comment:
You might not need the `StringOrFloat` helper or the visitor if you try to deserialize into a f64 first and then try the string:

```rust
match f64::deserialize(deserializer) {
    Ok(v) => return v,
    Err(_) => /* run existing string parsing code here */
}
```

Reply:
I think this doesn't work because the first call to `deserialize` consumes the `deserializer`. 
I could replace the visitor with `#[serde(untagged)]` though